README.md
# react-google-map ![npm](https://img.shields.io/npm/v/react-google-map.svg) ![license](https://img.shields.io/npm/l/react-google-map.svg)

React component to render a Google Map with markers.
You can use all official Google Maps API features.

https://developers.google.com/maps/documentation/javascript/reference

![react-google-map example](/screenshots/react-google-map-exemple.png)

## Install

```sh
npm install --save react-google-map
```

__If you don't have a solution to load `googleMaps`, you could use this package:__

```sh
npm install --save react-google-maps-loader
```

## Changelog

See [changelog](./CHANGELOG.md)

## Demo

<http://xuopled.github.io/react-google-map/>

## Usage
```css

.map {
  height: 300px;
}

@media screen and (min-width: 1024px){
    .map {
        height: 500px;
    }
}
```

```js
import React, {PropTypes} from "react"

import GoogleMap from "react-google-map"
import GoogleMapLoader from "react-google-maps-loader"

import iconMarker from "./assets/iconMarker.svg"
import iconMarkerHover from "./assets/iconMarkerHover.svg"

import styles from "./index.css"

const MY_API_KEY = "AIzaSyDwsdjfskhdbfjsdjbfksiTgnoriOAoUOgsUqOs10J0" // fake

const Map = ({googleMaps}) => (
  // GoogleMap component has a 100% height style.
  // You have to set the DOM parent height.
  // So you can perfectly handle responsive with differents heights.
  <div className={styles.map}>
    <GoogleMap
      googleMaps={googleMaps}
      // You can add and remove coordinates on the fly.
      // The map will rerender new markers and remove the old ones.
      coordinates={[
        {
          title: "Toulouse",
          position: {
            lat: 43.604363,
            lng: 1.443363,
          },
          onLoaded: (googleMaps, map, marker) => {
            // Set Marker animation
            marker.setAnimation(googleMaps.Animation.BOUNCE)

            // Define Marker InfoWindow
            const infoWindow = new googleMaps.InfoWindow({
              content: `
                <div>
                  <h3>Toulouse<h3>
                  <div>
                    Toulouse is the capital city of the southwestern
                    French department of Haute-Garonne,
                    as well as of the Occitanie region.
                  </div>
                </div>
              `,
            })

            // Open InfoWindow when Marker will be clicked
            googleMaps.event.addListener(marker, "click", () => {
              infoWindow.open(map, marker)
            })

            // Change icon when Marker will be hovered
            googleMaps.event.addListener(marker, "mouseover", () => {
              marker.setIcon(iconMarkerHover)
            })

            googleMaps.event.addListener(marker, "mouseout", () => {
              marker.setIcon(iconMarker)
            })

            // Open InfoWindow directly
            infoWindow.open(map, marker)
          },
        }
      ]}
      center={{lat: 43.604363, lng: 1.443363}}
      zoom={8}
      onLoaded={(googleMaps, map) => {
        map.setMapTypeId(googleMaps.MapTypeId.SATELLITE)
      }}
    />
  </div>
)

Map.propTypes = {
  googleMaps: PropTypes.object.isRequired,
}

export default GoogleMapLoader(Map, {
  libraries: ["places"],
  key: MY_API_KEY,
})

```

## Props
  * `autoFitBounds`: Boolean - Enable it if you will add and remove markers on the fly. Bounds will fit automatically
  * `boundsOffset`: Number - If `autoFitBounds` enabled you want custom bounds, - by default is 0.002
  * `coordinates`: Array of Marker props. You can use all props defined in `google.maps.MarkerOptions` object specification (https://developers.google.com/maps/documentation/javascript/reference#MarkerOptions). If you need some `google.maps` constants, use the `onLoaded` prop (`onLoaded: (googleMaps, map, marker) => {}`) to update your map and markers - by default is []
  * `googleMaps`: Object - injected by placesLoader,
  * `onLoaded`: Function with two parameters (`onLoaded: (googleMaps, map) => {}`),

You can use all props defined in `google.maps.MapOptions` object specification:
https://developers.google.com/maps/documentation/javascript/reference#MapOptions

If you need some `google.maps` constants, use the `onLoaded` prop

## Development

### Clean `lib` folder

```js
npm run clean
```

### Build `lib` folder

```js
npm run build
```

### Watch `src` folder

```js
npm run watch
```

### Lint `src` folder

```js
npm run lint
```

## License

See [MIT](./LICENCE)


CHANGELOG.md
# 3.1.1 - 2018-01-16

* Fixed : shouldComponentUpdate method

# 3.1.0 - 2017-09-28

* Update : dependencies

# 3.0.1 - 2016-11-20

* Remove : useless `react-google-maps-loader` dependency in package.json

# 3.0.0 - 2016-11-20

* Add : `onLoaded` prop on `map`
* Add : `onLoaded` prop on `marker`
* Update : custom props -> google maps native props

# 2.0.0 - 2016-11-18

* Update : package.json
* Update : .babelrc
* Update : .editorconfig
* Update : .eslinted
* Update : README
* Update : webpack.config.js
* Update : src/index.js - Add all google maps options

# 1.0.1 - 2015-09-11

* Add : libraryTarget into Webpack config

# 1.0.0 - 2015-09-09

* Initial release


src/demo.js
import './demo.css'
import 'react-google-places-suggest/lib/index.css'
import 'react-google-map/lib/index.css'

import React, { Component } from 'react'
import GooglePlacesSuggest from 'react-google-places-suggest'
import GoogleMap from 'react-google-map'

export default class Demo extends Component {
  state = {
    coordinates: [],
    search: '',
    selectedCoordinate: null,
    suggests: [],
    zoom: 8,
  }

  handleSearchChange = (e) => {
    this.setState({ search: e.target.value })
  }

  handleDescriptionChange = (e) => {
    this.setState({ description: e.target.value })
  }

  handleSelectSuggest = (suggestName, coordinate) => {
    this.setState({ search: suggestName, selectedCoordinate: coordinate })
  }

  handleClickOnAddPlace = () => {
    const { selectedCoordinate } = this.state
    this.addCoordinate(selectedCoordinate)
    this.setState({ search: '', selectedCoordinate: null })
  }

  handleRemoveCoordinate = (index) => {
    const { coordinates } = this.state
    const newCoordinates = coordinates.concat()
    newCoordinates.splice(index, 1)
    this.setState({ coordinates: newCoordinates })
  }

  handleZoomChange = (zoom) => {
    this.setState({ zoom: zoom })
  }

  handleMapChange = (coordinates, zoom) => {
    this.setState({ coordinates, zoom })
  }

  addCoordinate(coordinate) {
    const coordinates = [...this.state.coordinates, coordinate]
    this.setState({ coordinates })
  }

  renderSearch() {
    const { search } = this.state
    return (
      <div className="search">
        <GooglePlacesSuggest onSelectSuggest={ this.handleSelectSuggest } search={ search }>
          <input
            type="text"
            value={ search }
            placeholder="Search a location"
            onChange={ this.handleSearchChange }
          />
        </GooglePlacesSuggest>
      </div>
    )
  }

  renderForm() {
    const { selectedCoordinate } = this.state

    return (
      <div className="form">
        { this.renderSearch() }
        <span
          type="button"
          className="button"
          theme="blue"
          disabled={ !selectedCoordinate }
          onClick={ this.handleClickOnAddPlace }
        >
          Add marker
        </span>
      </div>
    )
  }

  renderMap() {
    const { zoom } = this.props
    const { coordinates } = this.state

    return (
      <GoogleMap
        coordinates={ coordinates }
        zoom={ zoom }
        onChange={ this.handleMapChange }
      />
    )
  }

  render() {
    return (
      <div className="container">
        <div>{ this.renderSearch() }</div>
        <div>{ this.renderMap() }</div>
      </div>
    )
  }
}


src/index.js
import PropTypes from "prop-types"
import React from "react"

const inlineStyles = {
  position: "relative",
  overflow: "hidden",
  height: "100%",
  width: "100%",
}

class GoogleMap extends React.Component {
  constructor() {
    super()
    this.state = {
      map: null,
      markers: new Map(),
    }
  }

  componentDidMount() {
    const {coordinates, googleMaps, onLoaded, ...props} = this.props

    const map = new googleMaps.Map(this.ref_map, {
      ...props,
    })

    this.setState({map}, () => {
      this.addNewMarkers(coordinates)
    })

    if (onLoaded) {
      onLoaded(googleMaps, map)
    }
  }

  componentWillReceiveProps(nextProps) {
    const newMarkers = nextProps.coordinates.some(
      coordinate => !this.state.markers.has(this.getMarkerId(coordinate))
    )
    const oldMarkers = [...this.state.markers.keys()].some(
      markerId =>
        !nextProps.coordinates.some(
          coordinate => markerId === this.getMarkerId(coordinate)
        )
    )

    if (oldMarkers) {
      this.removeOldMarkers(nextProps.coordinates)
    }

    if (newMarkers) {
      this.addNewMarkers(nextProps.coordinates)
    }
  }

  shouldComponentUpdate(nextProps, nextState) {
    return JSON.stringify(this.props.coordinates) !== JSON.stringify(nextProps.coordinates)
  }

  getMarkerId(coordinate) {
    return coordinate.position.lat + "_" + coordinate.position.lng
  }

  removeOldMarkers(coordinates) {
    const {markers} = this.state
    const {autoFitBounds} = this.props

    markers.forEach((marker, markerId) => {
      const isMarkerUsed = coordinates.some(
        coordinate => this.getMarkerId(coordinate) === markerId
      )

      if (!isMarkerUsed) {
        marker.setMap(null)
        markers.delete(markerId)
      }
    })

    this.setState({markers})

    if (autoFitBounds) {
      this.fitBounds()
    }
  }

  addNewMarkers(coordinates) {
    const {markers} = this.state
    const {autoFitBounds} = this.props

    coordinates.forEach(coordinate => {
      const markerId = this.getMarkerId(coordinate)
      if (!markers.has(markerId)) {
        markers.set(markerId, this.addMarker(markerId, coordinate))
      }
    })

    this.setState({markers})

    if (autoFitBounds) {
      this.fitBounds()
    }
  }

  addMarker(markerId, coordinate) {
    const {map} = this.state
    const {googleMaps} = this.props
    const {onLoaded, ...markerProps} = coordinate

    const marker = new googleMaps.Marker({
      map: map,
      ...markerProps,
    })

    if (onLoaded) {
      onLoaded(googleMaps, map, marker)
    }

    return marker
  }

  fitBounds() {
    const {map, markers} = this.state
    const {boundsOffset, googleMaps} = this.props

    if (!map || markers.size === 0) {
      return
    }

    const bounds = Array.from(markers.values()).reduce(
      (bound, marker) => bound.extend(marker.getPosition()),
      new googleMaps.LatLngBounds()
    )
    const center = bounds.getCenter()

    bounds
      .extend(
        new googleMaps.LatLng(
          center.lat() + boundsOffset,
          center.lng() + boundsOffset
        )
      )
      .extend(
        new googleMaps.LatLng(
          center.lat() - boundsOffset,
          center.lng() - boundsOffset
        )
      )

    map.setCenter(center)
    map.fitBounds(bounds)
  }

  render() {
    return <div ref={ref => (this.ref_map = ref)} style={inlineStyles} />
  }
}

GoogleMap.propTypes = {
  autoFitBounds: PropTypes.bool,
  boundsOffset: PropTypes.number,
  coordinates: PropTypes.arrayOf(
    PropTypes.shape({
      onLoaded: PropTypes.func,
    })
  ),
  googleMaps: PropTypes.object.isRequired,
  onLoaded: PropTypes.func,
}

GoogleMap.defaultProps = {
  autoFitBounds: false,
  boundsOffset: 0.002,
  coordinates: [],
  onLoaded: null,
}

export default GoogleMap


webpack.config.js
var webpack = require("webpack")
var path = require("path")
var ExtractTextPlugin = require('extract-text-webpack-plugin')

module.exports = {
  debug: true,
  devtool: "#eval-source-map",
  entry: "./src/entry.js",
  output: {
    path: __dirname,
    filename: "./build/bundle.js",
  },
  resolve: {
    extensions: ["", ".js", ".css"]
  },
  module: {
    preLoaders: [
      { test: /\.js?$/, exclude: /node_modules/, loader: "eslint" },
    ],
    loaders: [
      { test: /\.js$/, exclude: /node_modules/, loader: "babel" },
      { test: /\.css$/, loader: "style!css" },
      { test: /\.svg$/, loader: "file?name=build/[name].[ext]" },
    ],
  },
  plugins: [
    new webpack.NoErrorsPlugin(),
    new webpack.optimize.OccurenceOrderPlugin(),
    new webpack.optimize.DedupePlugin(),
    new webpack.optimize.AggressiveMergingPlugin(),
    new webpack.optimize.UglifyJsPlugin({
      compress: {
        warnings: false,
      },
    }),
    new ExtractTextPlugin('./build/bundle.css'),
  ]
}


src/Map.js
import React, {Component} from "react"

import GoogleMap from "react-google-map"
import GoogleMapLoader from "react-google-maps-loader"

import iconMarker from "./assets/iconMarker.svg"
import iconMarkerHover from "./assets/iconMarkerHover.svg"

class Map extends Component {
  render() {
    const {googleMaps} = this.props
    return (
      <div style={{height: "500px"}}>
        <GoogleMap
          googleMaps={googleMaps}
          coordinates={[
            {
              icon: iconMarker,
              title: "Toulouse",
              position: {
                lat: 43.604363,
                lng: 1.443363,
              },
              onLoaded: (googleMaps, map, marker) => {
                // Set Marker animation
                marker.setAnimation(googleMaps.Animation.BOUNCE)

                // Define Marker InfoWindow
                const infoWindow = new googleMaps.InfoWindow({
                  content: `
                    <div>
                      <h3>Toulouse</h3>
                      <div>
                        Toulouse is the capital city of the southwestern
                        French department of Haute-Garonne,
                        as well as of the Occitanie region.
                      </div>
                    </div>
                  `,
                })

                // Open InfoWindow when Marker will be clicked
                googleMaps.event.addListener(marker, "click", () => {
                  infoWindow.open(map, marker)
                })

                // Change icon when Marker will be hovered
                googleMaps.event.addListener(marker, "mouseover", () => {
                  marker.setIcon(iconMarkerHover)
                })

                googleMaps.event.addListener(marker, "mouseout", () => {
                  marker.setIcon(iconMarker)
                })

                // Open InfoWindow directly
                infoWindow.open(map, marker)
              },
            },
          ]}
          center={{lat: 43.604363, lng: 1.443363}}
          zoom={8}
          onLoaded={(googleMaps, map) => {
            map.setMapTypeId(googleMaps.MapTypeId.HYBRID)
          }}
        />
      </div>
    )
  }
}

export default GoogleMapLoader(Map, {
  libraries: ["places"],
  key: "AIzaSyAGz5ACyQ8hyEB-foU68UYgR66ISFJ8ofI",
})
